### PR TITLE
Expand entity-filter options

### LINKED ISSUE #3543: Entity Filter - Attributes Filter [Feature Request]
*@simonparkis commented on Mar 6, 2019, 9:54 PM UTC:*

It would be very useful to be able to use the entity-filter card to filter attributes as well as states, specifically with fine controls.

The example that comes to mind would be to filter items in a list of battery powered devices, with which the battery attribute is reporting less than 10%. This would allow to make a card with devices requiring battery replacement.

*This issue was moved by [iantrich](https://github.com/iantrich) from [home-assistant/ui-schema#241](https://github.com/home-assistant/ui-schema/issues/241).*

### PR DESCRIPTION
**Pull request by iantrich**
Closes https://github.com/home-assistant/home-assistant-polymer/issues/3543

Adds the following to `state_filter`:
- `operator` option. Specify operator to use in comparison. Options are `==, >, >=, <, <=, !=, regex`
- `value` option. Specify value to compare against.
- `attribute` option.  Specify attribute to use instead of state in comparison.

Adds `state_filter` as an option to entities as well

e.g.
```yaml
type: entity-filter
state_filter:
  - "on"
  - operator: ">"
    value: 90
entities:
  - sensor.water_leak
  - sensor.outside_temp
  - entity: sensor.humidity_and_temp
    state_filter:
      - operator: ">"
        value: 50
        attribute: humidity
```

### CODE REVIEW

**Comment by iantrich** [comment]
Need to make a change to setConfig before this is reviewed/merged

File: src/panels/lovelace/common/evaluate-filter.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { HassEntity } from "home-assistant-js-websocket";
+
+export const evaluateFilter = (stateObj: HassEntity, filter: any): boolean => {
+  const operator = filter.operator ? filter.operator : "==";
```
Comment (bramkragten) [inline src/panels/lovelace/common/evaluate-filter.ts]:
```suggestion
  const operator = filter.operator || "==";
```

File: src/panels/lovelace/common/evaluate-filter.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  const value = filter.value ? filter.value : filter;
```
Comment (bramkragten) [inline src/panels/lovelace/common/evaluate-filter.ts]:
```suggestion
  const value = filter.value || filter;
```

File: src/panels/lovelace/cards/hui-entity-filter-card.ts [hunk 3]
```diff
@@ -50,13 +62,36 @@ class EntityFilterCard extends HTMLElement implements LovelaceCard {
 
     this._hass = hass;
 
+    if (!this._config.entities) {
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-entity-filter-card.ts]: Shouldn't this error in `setConfig`?

**Comment by bramkragten** [comment]
I think your example is wrong, the `state_filter` of an entity should be an array?

File: src/panels/lovelace/cards/hui-entity-filter-card.ts [hunk 4]
```diff
@@ -2,25 +2,37 @@ import { createCardElement } from "../common/create-card-element";
...
+  private _oldEntities?: EntityFilterEntityConfig[];
 
   public getCardSize(): number {
     return this._element ? this._element.getCardSize() : 1;
   }
 
   public setConfig(config: EntityFilterCardConfig): void {
-    if (!config.state_filter || !Array.isArray(config.state_filter)) {
+    if (
+      (!config.state_filter &&
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-entity-filter-card.ts]: Wouldn't this do it?
```
     if ((!config.state_filter || !Array.isArray(config.state_filter)) &&
        config.entities.filter((entity) => {
          return Boolean(typeof entity === "object" && entity.state_filter && Array.isArray(entity.state_filter);
        }).length !== config.entities.length)
```

Comment (iantrich) [reply src/panels/lovelace/cards/hui-entity-filter-card.ts]: That would require that each entity have a state_filter

Comment (bramkragten) [reply src/panels/lovelace/cards/hui-entity-filter-card.ts]: No, it requires a default `state_filter` OR all entities have their own

Comment (iantrich) [reply src/panels/lovelace/cards/hui-entity-filter-card.ts]: I see it now. Thanks, that's much cleaner!

Comment (bramkragten) [reply src/panels/lovelace/cards/hui-entity-filter-card.ts]: Even better to use `every`:
```
if (
    !(config.state_filter && Array.isArray(config.state_filter)) ||
    !config.entities.every((entity) => typeof entity === "object" && entity.state_filter && Array.isArray(entity.state_filter)
)
```

File: src/panels/lovelace/cards/hui-entity-filter-card.ts [hunk 5]
```diff
@@ -36,6 +36,10 @@ class EntityFilterCard extends HTMLElement implements LovelaceCard {
       throw new Error("Incorrect filter config.");
     }
 
+    if (!config.entities) {
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-entity-filter-card.ts]: Place this check above the check for filter config, and check if it is an array, as we assume in that check entities is an array.

File: src/panels/lovelace/cards/hui-entity-filter-card.ts (hunk 4 above)
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-entity-filter-card.ts]: @iantrich

**Review by bramkragten — APPROVED** [review]

**Pull request merged**